### ISSUE TITLE
Question about energy function

### ISSUE THREAD

**Issue by Shi-qingyu** [Bujiazi/MotionClone#22]
Thank you for your outstanding work!
I've noticed that the guidance in the generation process is formulated as:
![Image](https://github.com/user-attachments/assets/c01d3654-f798-4192-9ccf-35a62b55f915)
where 𝑔 represents the MSE between the sparse temporal attention map of the reference video and the generated video, which we aim to minimize.
My concern is that this function is derived from the paper 'Diffusion Models Beat GANs on Image Synthesis.'
![Image](https://github.com/user-attachments/assets/b25d0829-d88f-4428-8774-2f92dac44194)
Where 𝑝 is a classifier that we aim to maximize. Is there a conflict between 𝑔 and 𝑝 due to their differing optimization target? Please correct me if I have any misunderstanding!

Best Regards,
Qingyu

**Comment by LPengYang** [comment]
Sorry for being late. Thanks for your attention on our work. In my opinion, during sampling for controllable generation, one can define customized functions to indicate the ideal target. This target, in different versions, is viewed to have different features, such as being classified as true; possessing high energy value, or exhibiting low loss value. **These all refer to the same target,** i.e., the ideal targeted generation.  For instance, the presented two equations refer to the same target because the sampling point with a low loss value should be classified as true if the classifier is well-designed. What matters is how to define the ideal target and we can accordingly define customized functions to sample towards it.

**Comment by LPengYang** [comment]
> Thank you for your outstanding work! I've noticed that the guidance in the generation process is formulated as: ![Image](https://github.com/user-attachments/assets/c01d3654-f798-4192-9ccf-35a62b55f915) where 𝑔 represents the MSE between the sparse temporal attention map of the reference video and the generated video, which we aim to minimize. My concern is that this function is derived from the paper 'Diffusion Models Beat GANs on Image Synthesis.' ![Image](https://github.com/user-attachments/assets/b25d0829-d88f-4428-8774-2f92dac44194) Where 𝑝 is a classifier that we aim to maximize. Is there a conflict between 𝑔 and 𝑝 due to their differing optimization target? Please correct me if I have any misunderstanding!
> 
> Best Regards, Qingyu

Sorry for being late. Thanks for your attention on our work. In my opinion, during sampling for controllable generation, one can define customized functions to indicate the ideal target. This target, in different versions, is viewed to have different features, such as being classified as true; possessing high energy value, or exhibiting low loss value. These all refer to the same target, i.e., the ideal targeted generation. For instance, the presented two equations refer to the same target because the sampling point with a low loss value should be classified as true if the classifier is well-designed. What matters is how to define the ideal target and we can accordingly define customized functions to sample towards it.

**Comment by LPengYang** [comment]
> Thank you for your outstanding work! I've noticed that the guidance in the generation process is formulated as: ![Image](https://github.com/user-attachments/assets/c01d3654-f798-4192-9ccf-35a62b55f915) where 𝑔 represents the MSE between the sparse temporal attention map of the reference video and the generated video, which we aim to minimize. My concern is that this function is derived from the paper 'Diffusion Models Beat GANs on Image Synthesis.' ![Image](https://github.com/user-attachments/assets/b25d0829-d88f-4428-8774-2f92dac44194) Where 𝑝 is a classifier that we aim to maximize. Is there a conflict between 𝑔 and 𝑝 due to their differing optimization target? Please correct me if I have any misunderstanding!
> 
> Best Regards, Qingyu

Sorry for being late. From the implementation of diffusion beats gan: [line1](https://github.com/openai/guided-diffusion/blob/22e0df8183507e13a7813f8d38d51b072ca1e67c/guided_diffusion/gaussian_diffusion.py#L384) and [line2](https://github.com/openai/guided-diffusion/blob/22e0df8183507e13a7813f8d38d51b072ca1e67c/scripts/classifier_sample.py#L63), it seems that we peform the same operation.

**Comment by Shi-qingyu** [comment]
Hi! Thank you for your response, and I am following this excellent work. I agree with you that the same operation in diffusion beats gan is performed. What confuses me is that in diffusion beats gan, this operation increases the log p(y|x) (classifier). Given that the same operation is applied, shouldn't g(z, y, t) (energy function which is MSE Loss) also increase, which seems unexpected?

Best Regards,
Qingyu

**Comment by Shi-qingyu** [comment]
Furthermore, I have run the code, and the generated videos look perfect. I’m just curious about the theoretical basis of this guidance operation.

**Comment by LPengYang** [comment]
> Hi! Thank you for your response, and I am following this excellent work. I agree with you that the same operation in diffusion beats gan is performed. What confuses me is that in diffusion beats gan, this operation increases the log p(y|x) (classifier). Given that the same operation is applied, shouldn't g(z, y, t) (energy function which is MSE Loss) also increase, which seems unexpected?
> 
> Best Regards, Qingyu

Hi~ It seems to me that the second equation refers to the score function, which indicates the direction toward the target (should be added). However, in practice, this is achieved by the gradient descent direction indicated by the loss function. They refer to the same target.

**Comment by LPengYang** [comment]
> Furthermore, I have run the code, and the generated videos look perfect. I’m just curious about the theoretical basis of this guidance operation.

Hi~ Sorry for the misunderstanding. We have identified the issue that causes this discrepancy. It seems that our implementation does not follow the standard form. In diffusion beats gan, the modified noise is used for both predicted $x_{0}$ and adding noise (The modified noise is negatively related with $x_{t-1}$).  However, the current implementation first uses vanilla noise to predict $x_{0}$ and then uses modified noise as adding noise (The modified noise is positively related with $x_{t-1}$, see [line](https://github.com/LPengYang/MotionClone/blob/a79773541a275e17acfec0194a5e2cd22acb54a4/motionclone/utils/motionclone_functions.py#L382)). Therefore, they refer to the same target. Fortunately, a translation between them can be achieved by modifying the coefficient, which is determined for pre-trained diffusion models. We will incorporate an appropriate update in the final version to address this. Sorry again for the misunderstanding.

**Comment by Shi-qingyu** [comment]
Thank you for the detailed explanation! It has resolved my confusion. Please feel free to close this issue. Thanks again for your thorough response.

**Comment by LPengYang** [comment]
> Thank you for the detailed explanation! It has resolved my confusion. Please feel free to close this issue. Thanks again for your thorough response.

Thanks for your careful review! It has been derived that $\lambda_1 = (\sqrt{1/\alpha_t} -1)\lambda_2$, in which $\lambda_1$ and $\lambda_2$ are the guidance weight in equation 1 and equation 2, respectively. It has been validated by using the guidance form in equation 2 with $\lambda_2 = \lambda_1/(\sqrt{1/\alpha_t} -1)$ and the same results can be obtained. Given that the experiments are conducted with the first form, we will update the equation to align with "diffusion beats gan", and use $\lambda_2 = \lambda_1/(\sqrt{1/\alpha_t} -1)$ in our code with a special explanation.

**Comment by Shi-qingyu** [comment]
> > Thank you for the detailed explanation! It has resolved my confusion. Please feel free to close this issue. Thanks again for your thorough response.
> 
> Thanks for your careful review! It has been derived that λ 1 = ( 1 / α t − 1 ) λ 2 , in which λ 1 and λ 2 are the guidance weight in equation 1 and equation 2, respectively. It has been validated by using the guidance form in equation 2 with λ 2 = λ 1 / ( 1 / α t − 1 ) and the same results can be obtained. Given that the experiments are conducted with the first form, we will update the equation to align with "diffusion beats gan", and use λ 2 = λ 1 / ( 1 / α t − 1 ) in our code with a special explanation.

I see. This is very clear!

**Comment by YingjieYin** [comment]
Hello, author!  how to get the relationship λ 2 = λ 1 / ( 1 / α t − 1 ), i have try to deduce the relationship ,but can't get the conclution λ 2 = λ 1 / ( 1 / α t − 1 )

**Comment by LPengYang** [comment]
> Hello, author! how to get the relationship λ 2 = λ 1 / ( 1 / α t − 1 ), i have try to deduce the relationship ,but can't get the conclution λ 2 = λ 1 / ( 1 / α t − 1 )

Hello, sorry for being late. The implementations corresponding to these two formulations are presented in https://github.com/LPengYang/MotionClone/blob/8b3d937f005a90c2950c2ec96332fa5c34190c70/motionclone/utils/motionclone_functions.py#L286 and https://github.com/LPengYang/MotionClone/blob/8b3d937f005a90c2950c2ec96332fa5c34190c70/motionclone/utils/motionclone_functions.py#L413
You can replace one with the other and obtain the same results.
